drivers/hwinfo/hwinfo_nrf.c

```diff
@@ -25,17 +25,34 @@ struct nrf_uid {
 ssize_t z_impl_hwinfo_get_device_id(uint8_t *buffer, size_t length)
 {
 	struct nrf_uid dev_id;
-	uint32_t deviceid[2];
+	uint32_t buf[2];
 
+#if NRF_FICR_HAS_DEVICE_ID || NRF_FICR_HAS_INFO_DEVICE_ID
+	/* DEVICEID is accessible, use this */
 #if defined(CONFIG_TRUSTED_EXECUTION_NONSECURE) && defined(NRF_FICR_S)
-	soc_secure_read_deviceid(deviceid);
+	soc_secure_read_deviceid(buf);
 #else
-	deviceid[0] = nrf_ficr_deviceid_get(NRF_FICR, 0);
-	deviceid[1] = nrf_ficr_deviceid_get(NRF_FICR, 1);
+	buf[0] = nrf_ficr_deviceid_get(NRF_FICR, 0);
+	buf[1] = nrf_ficr_deviceid_get(NRF_FICR, 1);
+#endif
+#elif NRF_FICR_HAS_DEVICE_ADDR || NRF_FICR_HAS_BLE_ADDR
+	/* DEVICEID is not accessible, use device/ble address instead.
+	 * Assume that it is always accessible from the non-secure image.
+	 */
+	buf[0] = nrf_ficr_deviceaddr_get(NRF_FICR, 0);
+	buf[1] = nrf_ficr_deviceaddr_get(NRF_FICR, 1);
+
+	/* Assume that ER and IR are available whenever deviceaddr is.
+	 * Use the LSBytes from ER and IR to complete the device id.
+	 */
+	buf[1] |= ((nrf_ficr_er_get(NRF_FICR, 0) & 0xFF) << 16);
+	buf[1] |= ((nrf_ficr_ir_get(NRF_FICR, 0) & 0xFF) << 24);
```

Comment:
I guess the BLE address is already unique, so what are these two numbers?

Reply:
The BLE address is only 48-bit, so the extra 16 bits are taken from ER and IR.

Reply:
Why not only returning 48 bits in this case?

Reply:
Because we have bytes enough that are ensured unique, so why not use them to actually get a full 64-bit identifier? It is arbitrary but as long as it does not change I see not issue with it.

Reply:
Do you have a description of this registers? We can have more than 64 bytes too. But for me 48 is enough because they are guaranteed to be unique.

Reply:
The registers are described in the datasheet, they are basically these:
https://infocenter.nordicsemi.com/index.jsp?topic=%2Fps_nrf52840%2Fficr.html&cp=5_0_0_3_3&anchor=register.ER-0-3
https://infocenter.nordicsemi.com/index.jsp?topic=%2Fps_nrf52840%2Fficr.html&cp=5_0_0_3_3&anchor=register.ER-0-3

The reason I want to fill 64-bit is that this is the size used in the API and the number of bits that we fill on other SoC families when using the `DEVICEID` register

Reply:
I did think about that. We could use ER or IR, since each is 128-bit long. But I'd argue that the BLE device address conceptually maps better to an identifier, so that's why I think it's a better choice. Then we are 2 bytes short, so I take them from ER and IR. I think this is reasonable.

Reply:
On the other hand, the current proposal requires 4 loads from the memory (probably crossing async bridges, so relatively slow), while using BLE address only would be 2 loads - 2 times faster. You can fill the remaining 16 bytes with arbitrary data like `0x00` or a part of Nordic's IEEE OUI (F4-CE-36).

But it's not a strong opinion. I expect this function to be used rarely, so the performance impact on the application is relatively low.

Reply:
This is obtaining a hw id, I do not think this will ever be called from a performance-sensitive code path.

Reply:
How about taking the last two bytes only from IR?

Reply:
I also thought about that but conceptually using all the unique keys on the device sounded more elegant. There's very little difference in practice.

Reply:
The thing ist that they are guaranteed unique only if you take the full number. No idea how nordic handles that. Since the BLE address is unique, I'm ok. Still think it's odd, but no blocker.

Reply:
USB Mass Storage Class – Bulk Only Transport 4.1.1 Serial Number
> The iSerialNumber field shall be set to the index of the string descriptor that contains the serial number. The
serial number shall contain at least 12 valid digits, represented as a UNICODE string. The last 12 digits of the
serial number shall be unique to each USB idVendor and idProduct pair.

Therefore 48 bits are enough for MSC. However, it is important that the 48-bit MAC address would appear on the last 12 digits (hexadecimal digits, so 12*4=48 bits).

To fulfill the MSC serial number requirements, the extra data from ER and IR should come at the beginning instead of the end.